src/engine.rs

```diff
@@ -365,49 +408,101 @@ impl Engine {
                 Command::Dispatch(shader_id, wg_size, bindings) => {
                     // println!("dispatching {:?} with {} bindings", wg_size, bindings.len());
                     let shader = &self.shaders[shader_id.0];
-                    let bind_group = self.bind_map.create_bind_group(
-                        device,
-                        &shader.bind_group_layout,
-                        bindings,
-                        external_resources,
-                        &mut self.pool,
-                    )?;
-                    let mut cpass = encoder.begin_compute_pass(&Default::default());
-                    #[cfg(feature = "wgpu-profiler")]
-                    profiler.begin_scope(shader.label, &mut cpass, device);
-                    cpass.set_pipeline(&shader.pipeline);
-                    cpass.set_bind_group(0, &bind_group, &[]);
-                    cpass.dispatch_workgroups(wg_size.0, wg_size.1, wg_size.2);
-                    #[cfg(feature = "wgpu-profiler")]
-                    profiler.end_scope(&mut cpass);
+                    if let Some(cpu_shader) = shader.cpu_shader {
+                        let resources =
+                            transient_map.create_cpu_resources(&mut self.bind_map, bindings);
+                        cpu_shader(wg_size.0, &resources);
```

Comment:
I recommend adding a brief comment here that explains that the stage will be executed on the CPU synchronously right here. I think it's worth explaining, in a comment, how you foresee this system will work when CPU and GPU execution of pipelines can be mixed in the future.

For example, AIUI as long as no CPU stage depends on a preceding GPU stage, all buffers that are used by downstream GPU pipelines are guaranteed to be fully initialized at the time of dispatch. I would add a note that it is (currently) forbidden for a CPU stage to depend on the output of a GPU stage because the necessary synchronization mechanism is not supported. I would add to the comment that this is not an issue because (currently) the CPU flag is all or nothing and prevents mixing of stages.